File: internal/installer/install/executable.go
package install

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/gvcgo/goutils/pkgs/gtea/gprint"
	"github.com/gvcgo/goutils/pkgs/gtea/spinner"
	"github.com/gvcgo/goutils/pkgs/gutils"
	"github.com/gvcgo/goutils/pkgs/request"
	"github.com/gvcgo/version-manager/internal/cnf"
	"github.com/gvcgo/version-manager/internal/download"
	"github.com/gvcgo/version-manager/internal/luapi/lua_global"
	"github.com/gvcgo/version-manager/internal/utils"
)

const (
	MinicondaSDKName string = "miniconda"
	ErlangSDKName    string = "erlang"
	ElixirSDKName    string = "elixir"
	VSCodeSDKName    string = "vscode"
)

/*
Install miniconda.
*/
func InstallMiniconda(exePath, installDir string) (task *utils.SysCommandRunner) {
	var commands []string
	homeDir, _ := os.UserHomeDir()
	if runtime.GOOS == "windows" {
		/*
			start /wait "" Miniconda3-latest-Windows-x86_64.exe /InstallationType=JustMe /RegisterPython=0 /S /D=%UserProfile%\Miniconda3
		*/
		commands = []string{
			"start",
			"/wait",
			"",
			exePath,
			"/InstallationType=JustMe",
			"/RegisterPython=0",
			"/S",
			fmt.Sprintf("/D=%s", installDir),
		}
	} else {
		gutils.ExecuteSysCommand(true, homeDir, "chmod", "+x", exePath)
		/*
			bash ~/miniconda.sh -b -p $HOME/miniconda
		*/
		commands = []string{
			"bash",
			exePath,
			"-b",
			"-p",
			installDir,
		}
	}
	// _, err = gutils.ExecuteSysCommand(true, homeDir, commands...)
	task = utils.NewSysCommandRunner(true, homeDir, commands...)
	return
}

// vscode
func InstallVSCode(pkgFilePath, installDir string) (task *utils.SysCommandRunner) {
	homeDir, _ := os.UserHomeDir()
	switch runtime.GOOS {
	case gutils.Windows:
		if strings.HasSuffix(pkgFilePath, ".exe") {
			// _, err = gutils.ExecuteSysCommand(true, homeDir,
			// 	pkgFilePath, "/VERYSILENT", "/MERGETASKS=!runcode")
			task = utils.NewSysCommandRunner(
				true,
				homeDir,
				pkgFilePath,
				"/VERYSILENT",
				"/MERGETASKS=!runcode",
			)
		}
	case gutils.Linux:
		if strings.HasSuffix(pkgFilePath, ".deb") {
			// _, err = gutils.ExecuteSysCommand(true, homeDir,
			// 	"sudo", "dpkg", "-i", pkgFilePath)
			task = utils.NewSysCommandRunner(
				true,
				homeDir,
				"sudo",
				"dpkg",
				"-i",
				pkgFilePath,
			)
		} else if strings.HasSuffix(pkgFilePath, ".rpm") {
			// _, err = gutils.ExecuteSysCommand(true, homeDir,
			// 	"sudo", "rpm", "-ivh", pkgFilePath)
			task = utils.NewSysCommandRunner(
				true,
				homeDir,
				"sudo",
				"rpm",
				"-ivh",
				pkgFilePath,
			)
		}
	case gutils.Darwin:
		err := utils.Extract(pkgFilePath, cnf.GetTempDir())
		if err != nil {
			gprint.PrintError("extract vscode failed: %v", err)
			os.RemoveAll(pkgFilePath)
			return
		}
		appName := "Visual Studio Code.app"
		ff := utils.NewFinder(appName)
		ff.Find(cnf.GetTempDir())
		appPath := filepath.Join(ff.GetDirName(), appName)
		if ok, _ := gutils.PathIsExist(appPath); ok {
			utils.MoveFileOnUnixSudo(appPath, "/Applications")
		}
		os.RemoveAll(cnf.GetTempDir())
	default:
	}
	return
}

/*
install *.exe for windows
1. erlang
2. elixir
*/
func InstallExeForWindows(exePath, installDir string) (task *utils.SysCommandRunner) {
	homeDir, _ := os.UserHomeDir()
	// _, err = gutils.ExecuteSysCommand(true, homeDir,
	// 	"start", "/wait", exePath, "/S", fmt.Sprintf("/D=%s", installDir))
	task = utils.NewSysCommandRunner(
		true,
		homeDir,
		"start",
		"/wait",
		exePath,
		"/S",
		fmt.Sprintf("/D=%s", installDir),
	)
	return
}

// Other standalone executables.
func InstallStandAloneExecutables(exePath, installDir string) (task *utils.SysCommandRunner) {
	fName := filepath.Base(exePath)
	os.MkdirAll(installDir, os.ModePerm)
	destPath := filepath.Join(installDir, fName)
	var err error
	if ok, _ := gutils.PathIsExist(exePath); ok {
		err = gutils.CopyAFile(exePath, destPath)
	}

	if err != nil {
		gprint.PrintError("%+v", err)
		os.RemoveAll(filepath.Dir(exePath))
	}

	if err == nil && runtime.GOOS != gutils.Windows {
		gutils.ExecuteSysCommand(true, installDir, "chmod", "+x", destPath)
	}
	return
}

/*
1. *.exe
2. *.deb
3. *.rpm
4. *.sh (miniconda)
5. unix-like executable
*/
type ExeInstaller struct {
	PluginName  string
	SDKName     string
	VersionName string
	Version     lua_global.Item
	Fetcher     *request.Fetcher
	spinner     *spinner.Spinner
	downloader  *download.Downloader
	installConf *lua_global.InstallerConfig
	signal      chan struct{}
}

func NewExeInstaller() (ei *ExeInstaller) {
	ei = &ExeInstaller{
		spinner:    spinner.NewSpinner(),
		downloader: download.NewDownloader(),
		signal:     make(chan struct{}),
	}
	return
}

func (ei *ExeInstaller) SetInstallConf(iconf *lua_global.InstallerConfig) {
	ei.installConf = iconf
}

func (ei *ExeInstaller) Initiate(pluginName, sdkName, versionName string, version lua_global.Item) {
	ei.PluginName = pluginName
	ei.SDKName = sdkName
	ei.VersionName = versionName
	ei.Version = version
}

func (ei *ExeInstaller) GetInstallDir() string {
	d := GetSDKVersionDir(ei.SDKName)
	return filepath.Join(d, fmt.Sprintf(VersionInstallDirPattern, ei.PluginName, ei.VersionName))
}

func (ei *ExeInstaller) GetSymbolLinkPath() string {
	d := GetSDKVersionDir(ei.SDKName)
	return filepath.Join(d, ei.SDKName)
}

func (ei *ExeInstaller) RenameFile() {
	if ei.installConf.BinaryRename != nil && ei.installConf.BinaryRename.NameFlag != "" {
		installDir := ei.GetInstallDir()
		dList, _ := os.ReadDir(installDir)
		for _, dd := range dList {
			if !dd.IsDir() && strings.Contains(dd.Name(), ei.installConf.BinaryRename.NameFlag) {
				newName := filepath.Join(installDir, ei.installConf.BinaryRename.RenameTo)
				if runtime.GOOS == "windows" {
					newName += ".exe"
				}
				os.Rename(filepath.Join(installDir, dd.Name()), newName)
			}
		}
	}
}

func (ei *ExeInstaller) Install() {
	localPath := ei.downloader.Download(ei.PluginName, ei.VersionName, ei.Version)
	if localPath == "" {
		return
	}
	var task *utils.SysCommandRunner

	switch ei.PluginName {
	case MinicondaSDKName:
		task = InstallMiniconda(localPath, ei.GetInstallDir())
	case ErlangSDKName, ElixirSDKName:
		task = InstallExeForWindows(localPath, ei.GetInstallDir())
	case VSCodeSDKName:
		task = InstallVSCode(localPath, ei.GetInstallDir())
	default:
		task = InstallStandAloneExecutables(localPath, ei.GetInstallDir())
		ei.RenameFile()
	}

	ei.spinner.SetTitle(fmt.Sprintf("Installing %s", ei.PluginName))
	ei.spinner.SetSweepFunc(func() {
		if task != nil {
			task.Cancel()
		}
		ei.signal <- struct{}{}
		os.RemoveAll(ei.GetInstallDir())
	})
	go ei.spinner.Run()

	go func() {
		var err error

		// switch ei.OriginSDKName {
		// case MinicondaSDKName:
		// 	err = InstallMiniconda(localPath, ei.GetInstallDir())
		// case ErlangSDKName, ElixirSDKName:
		// 	err = InstallExeForWindows(localPath, ei.GetInstallDir())
		// case VSCodeSDKName:
		// 	err = InstallVSCode(localPath, ei.GetInstallDir())
		// default:
		// 	err = InstallStandAloneExecutables(localPath, ei.GetInstallDir())
		// 	ei.RenameFile()
		// }

		if task != nil {
			err = task.Run()
		}
		if err != nil {
			gprint.PrintError("%+v", err)
			os.RemoveAll(filepath.Dir(localPath))
		}

		ei.signal <- struct{}{}
		ei.spinner.Quit()
	}()

	<-ei.signal
	// time.Sleep(time.Second * 2) // cursor fix.
}

File: rust-vmr/crates/vmr-install/src/executable.rs
use std::path::PathBuf;

use crate::common::{self, InstallerConfig};

// ---------------------------------------------------------------------------
// ExeInstaller — install SDKs distributed as executables / installers
// ---------------------------------------------------------------------------

/// Handles various executable-based installation strategies:
/// - Standalone executables (copy + chmod)
/// - Miniconda shell installers
/// - Visual Studio Code (platform-specific packages)
/// - Windows EXE installers (Erlang, Elixir, etc.)
pub struct ExeInstaller {
    pub plugin_name: String,
    pub sdk_name: String,
    pub version_name: String,
    pub version: vmr_download::Item,
    pub install_conf: Option<InstallerConfig>,
}

impl ExeInstaller {
    pub fn new() -> Self {
        ExeInstaller {
            plugin_name: String::new(),
            sdk_name: String::new(),
            version_name: String::new(),
            version: vmr_download::Item {
                url: String::new(),
                arch: String::new(),
                os: String::new(),
                installer: String::new(),
                sum: String::new(),
                sum_type: String::new(),
                size: 0,
            },
            install_conf: None,
        }
    }

    /// Returns the installation directory:
    /// `{sdk_version_dir}/{plugin_name}-{version_name}`
    pub fn get_install_dir(&self) -> PathBuf {
        let d = common::get_sdk_version_dir(&self.sdk_name);
        d.join(format!("{}-{}", self.plugin_name, self.version_name))
    }

    /// Returns the symlink path: `{sdk_version_dir}/{sdk_name}`
    pub fn get_symbol_link_path(&self) -> PathBuf {
        let d = common::get_sdk_version_dir(&self.sdk_name);
        d.join(&self.sdk_name)
    }

    /// After install, rename binaries according to `BinaryRename` config.
    fn rename_files(&self) {
        if let Some(ref conf) = self.install_conf {
            if let Some(ref br) = conf.binary_rename {
                if br.name_flag.is_empty() {
                    return;
                }
                let install_dir = self.get_install_dir();
                if let Ok(entries) = std::fs::read_dir(&install_dir) {
                    for entry in entries.flatten() {
                        let ft = entry.file_type().map(|t| t.is_dir()).unwrap_or(true);
                        if ft {
                            continue;
                        }
                        let name = entry.file_name().to_string_lossy().to_string();
                        if name.contains(&br.name_flag) {
                            let mut new_name = br.rename_to.clone();
                            if cfg!(target_os = "windows") {
                                new_name.push_str(".exe");
                            }
                            let new_path = install_dir.join(&new_name);
                            let _ = std::fs::rename(entry.path(), &new_path);
                        }
                    }
                }
            }
        }
    }

    // -----------------------------------------------------------------------
    // Install strategies (private helpers)
    // -----------------------------------------------------------------------

    /// Install a .sh-based Miniconda installer.
    ///
    /// Linux/macOS: `bash <exePath> -b -p <installDir>`
    /// Windows: `start /wait "" <exePath> /InstallationType=JustMe /RegisterPython=0 /S /D=<installDir>`
    fn install_miniconda(&self, exe_path: &str, install_dir: &str) {
        let home_dir = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));

        if cfg!(target_os = "windows") {
            let _ = std::process::Command::new("cmd")
                .args([
                    "/C",
                    "start",
                    "/wait",
                    "",
                    exe_path,
                    "/InstallationType=JustMe",
                    "/RegisterPython=0",
                    "/S",
                    &format!("/D={}", install_dir),
                ])
                .current_dir(&home_dir)
                .status();
        } else {
            // Make executable first
            #[cfg(unix)]
            {
                use std::os::unix::fs::PermissionsExt;
                let _ = std::fs::set_permissions(
                    std::path::Path::new(exe_path),
                    std::fs::Permissions::from_mode(0o755),
                );
            }
            let _ = std::process::Command::new("bash")
                .args([exe_path, "-b", "-p", install_dir])
                .current_dir(&home_dir)
                .status();
        }
    }

    /// Install a Windows .exe silently (Erlang / Elixir).
    ///
    /// `start /wait <exePath> /S /D=<installDir>`
    fn install_exe_for_windows(&self, exe_path: &str, install_dir: &str) {
        if !cfg!(target_os = "windows") {
            return;
        }
        let home_dir = dirs::home_dir().unwrap_or_else(|| PathBuf::from("C:\\"));
        let _ = std::process::Command::new("cmd")
            .args([
                "/C",
                "start",
                "/wait",
                exe_path,
                "/S",
                &format!("/D={}", install_dir),
            ])
            .current_dir(&home_dir)
            .status();
    }

    /// Install Visual Studio Code.
    ///
    /// - macOS: extract .zip, locate `Visual Studio Code.app`, `sudo mv` to
    ///   `/Applications`.
    /// - Linux: `sudo dpkg -i <pkg>` or `sudo rpm -ivh <pkg>`.
    /// - Windows: run `.exe` installer with `/VERYSILENT`.
    fn install_vscode(&self, pkg_path: &str, _install_dir: &str) {
        let home_dir = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));
        let pkg = std::path::Path::new(pkg_path);

        if cfg!(target_os = "macos") {
            let temp_dir = vmr_config::paths::get_temp_dir();
            if let Err(e) = vmr_utils::archive::extract(pkg, &temp_dir) {
                eprintln!("[exe] extract vscode failed: {}", e);
                let _ = std::fs::remove_file(pkg);
                return;
            }

            // Locate the .app bundle
            let app_name = "Visual Studio Code.app";
            let mut finder = vmr_utils::fs::HomeDirFinder::new(vec![app_name.to_string()]);
            finder.find(&temp_dir);
            if let Some(dir_name) = finder.get_dir_name() {
                let app_path = dir_name.join(app_name);
                if app_path.exists() {
                    // sudo mv to /Applications
                    let _ = std::process::Command::new("sudo")
                        .args(["mv", &app_path.to_string_lossy(), "/Applications/"])
                        .status();
                }
            }
            let _ = std::fs::remove_dir_all(&temp_dir);
        } else if cfg!(target_os = "linux") {
            let fname = pkg
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("")
                .to_lowercase();

            if fname.ends_with(".deb") {
                let _ = std::process::Command::new("sudo")
                    .args(["dpkg", "-i", pkg_path])
                    .current_dir(&home_dir)
                    .status();
            } else if fname.ends_with(".rpm") {
                let _ = std::process::Command::new("sudo")
                    .args(["rpm", "-ivh", pkg_path])
                    .current_dir(&home_dir)
                    .status();
            }
        } else if cfg!(target_os = "windows") {
            let fname = pkg
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("")
                .to_lowercase();
            if fname.ends_with(".exe") {
                let _ = std::process::Command::new(pkg_path)
                    .args(["/VERYSILENT", "/MERGETASKS=!runcode"])
                    .current_dir(&home_dir)
                    .status();
            }
        }
    }

    /// Install a standalone executable — copy to install dir and make it
    /// executable on Unix.
    fn install_standalone(&self, exe_path: &str) {
        let src = std::path::Path::new(exe_path);
        let fname = src
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("vmr_executable");

        let install_dir = self.get_install_dir();
        let _ = std::fs::create_dir_all(&install_dir);
        let dest_path = install_dir.join(fname);

        if src.exists() {
            if let Err(e) = std::fs::copy(src, &dest_path) {
                eprintln!("[exe] copy standalone executable failed: {}", e);
                if let Some(parent) = src.parent() {
                    let _ = std::fs::remove_dir_all(parent);
                }
                return;
            }
            // Make executable on Unix
            if !cfg!(target_os = "windows") {
                #[cfg(unix)]
                {
                    use std::os::unix::fs::PermissionsExt;
                    let _ = std::fs::set_permissions(
                        &dest_path,
                        std::fs::Permissions::from_mode(0o755),
                    );
                }
                #[cfg(not(unix))]
                let _ = ();
            }
        }
    }

    // -----------------------------------------------------------------------
    // Main install entry point
    // -----------------------------------------------------------------------

    /// Downloads the file and dispatches to the appropriate install strategy
    /// based on `plugin_name`.
    pub fn install(&mut self) {
        if self.version.url.is_empty() {
            eprintln!("[exe] empty download URL, skipping");
            return;
        }

        let mut dd = vmr_download::Downloader::new();
        let fpath = dd.download(&self.plugin_name, &self.version_name, self.version.clone());
        if fpath.is_empty() {
            eprintln!(
                "[exe] download failed for {}@{}",
                self.plugin_name, self.version_name
            );
            return;
        }

        let install_dir = self.get_install_dir();
        let install_dir_str = install_dir.to_string_lossy().to_string();

        match self.plugin_name.as_str() {
            common::MINICONDA_SDK_NAME => {
                self.install_miniconda(&fpath, &install_dir_str);
            }
            common::ERLANG_SDK_NAME | common::ELIXIR_SDK_NAME => {
                self.install_exe_for_windows(&fpath, &install_dir_str);
            }
            common::VSCODE_SDK_NAME => {
                self.install_vscode(&fpath, &install_dir_str);
            }
            _ => {
                self.install_standalone(&fpath);
                self.rename_files();
            }
        }
    }
}

impl Default for ExeInstaller {
    fn default() -> Self {
        Self::new()
    }
}
